File: models/v1/recipe_step.go
package models

import (
	"context"
	"net/http"
)

type (
	// RecipeStep represents a recipe step.
	RecipeStep struct {
		ID                        uint64  `json:"id"`
		Index                     uint    `json:"index"`
		PreparationID             uint64  `json:"preparationID"`
		PrerequisiteStep          uint64  `json:"prerequisiteStep"`
		MinEstimatedTimeInSeconds uint32  `json:"minEstimatedTimeInSeconds"`
		MaxEstimatedTimeInSeconds uint32  `json:"maxEstimatedTimeInSeconds"`
		TemperatureInCelsius      *uint16 `json:"temperatureInCelsius"`
		Notes                     string  `json:"notes"`
		RecipeID                  uint64  `json:"recipeID"`
		CreatedOn                 uint64  `json:"createdOn"`
		LastUpdatedOn             *uint64 `json:"lastUpdatedOn"`
		ArchivedOn                *uint64 `json:"archivedOn"`
		BelongsToRecipe           uint64  `json:"belongsToRecipe"`
	}

	// RecipeStepList represents a list of recipe steps.
	RecipeStepList struct {
		Pagination
		RecipeSteps []RecipeStep `json:"recipe_steps"`
	}

	// RecipeStepCreationInput represents what a user could set as input for creating recipe steps.
	RecipeStepCreationInput struct {
		Index                     uint    `json:"index"`
		PreparationID             uint64  `json:"preparationID"`
		PrerequisiteStep          uint64  `json:"prerequisiteStep"`
		MinEstimatedTimeInSeconds uint32  `json:"minEstimatedTimeInSeconds"`
		MaxEstimatedTimeInSeconds uint32  `json:"maxEstimatedTimeInSeconds"`
		TemperatureInCelsius      *uint16 `json:"temperatureInCelsius"`
		Notes                     string  `json:"notes"`
		RecipeID                  uint64  `json:"recipeID"`
		BelongsToRecipe           uint64  `json:"-"`
	}

	// RecipeStepUpdateInput represents what a user could set as input for updating recipe steps.
	RecipeStepUpdateInput struct {
		Index                     uint    `json:"index"`
		PreparationID             uint64  `json:"preparationID"`
		PrerequisiteStep          uint64  `json:"prerequisiteStep"`
		MinEstimatedTimeInSeconds uint32  `json:"minEstimatedTimeInSeconds"`
		MaxEstimatedTimeInSeconds uint32  `json:"maxEstimatedTimeInSeconds"`
		TemperatureInCelsius      *uint16 `json:"temperatureInCelsius"`
		Notes                     string  `json:"notes"`
		RecipeID                  uint64  `json:"recipeID"`
		BelongsToRecipe           uint64  `json:"belongsToRecipe"`
	}

	// RecipeStepDataManager describes a structure capable of storing recipe steps permanently.
	RecipeStepDataManager interface {
		RecipeStepExists(ctx context.Context, recipeID, recipeStepID uint64) (bool, error)
		GetRecipeStep(ctx context.Context, recipeID, recipeStepID uint64) (*RecipeStep, error)
		GetAllRecipeStepsCount(ctx context.Context) (uint64, error)
		GetAllRecipeSteps(ctx context.Context, resultChannel chan []RecipeStep) error
		GetRecipeSteps(ctx context.Context, recipeID uint64, filter *QueryFilter) (*RecipeStepList, error)
		GetRecipeStepsWithIDs(ctx context.Context, recipeID uint64, limit uint8, ids []uint64) ([]RecipeStep, error)
		CreateRecipeStep(ctx context.Context, input *RecipeStepCreationInput) (*RecipeStep, error)
		UpdateRecipeStep(ctx context.Context, updated *RecipeStep) error
		ArchiveRecipeStep(ctx context.Context, recipeID, recipeStepID uint64) error
	}

	// RecipeStepDataServer describes a structure capable of serving traffic related to recipe steps.
	RecipeStepDataServer interface {
		CreationInputMiddleware(next http.Handler) http.Handler
		UpdateInputMiddleware(next http.Handler) http.Handler

		ListHandler(res http.ResponseWriter, req *http.Request)
		CreateHandler(res http.ResponseWriter, req *http.Request)
		ExistenceHandler(res http.ResponseWriter, req *http.Request)
		ReadHandler(res http.ResponseWriter, req *http.Request)
		UpdateHandler(res http.ResponseWriter, req *http.Request)
		ArchiveHandler(res http.ResponseWriter, req *http.Request)
	}
)

// Update merges an RecipeStepInput with a recipe step.
func (x *RecipeStep) Update(input *RecipeStepUpdateInput) {
	if input.Index != x.Index {
		x.Index = input.Index
	}

	if input.PreparationID != x.PreparationID {
		x.PreparationID = input.PreparationID
	}

	if input.PrerequisiteStep != x.PrerequisiteStep {
		x.PrerequisiteStep = input.PrerequisiteStep
	}

	if input.MinEstimatedTimeInSeconds != x.MinEstimatedTimeInSeconds {
		x.MinEstimatedTimeInSeconds = input.MinEstimatedTimeInSeconds
	}

	if input.MaxEstimatedTimeInSeconds != x.MaxEstimatedTimeInSeconds {
		x.MaxEstimatedTimeInSeconds = input.MaxEstimatedTimeInSeconds
	}

	if input.TemperatureInCelsius != nil && input.TemperatureInCelsius != x.TemperatureInCelsius {
		x.TemperatureInCelsius = input.TemperatureInCelsius
	}

	if input.Notes != "" && input.Notes != x.Notes {
		x.Notes = input.Notes
	}

	if input.RecipeID != x.RecipeID {
		x.RecipeID = input.RecipeID
	}
}

// ToUpdateInput creates a RecipeStepUpdateInput struct for a recipe step.
func (x *RecipeStep) ToUpdateInput() *RecipeStepUpdateInput {
	return &RecipeStepUpdateInput{
		Index:                     x.Index,
		PreparationID:             x.PreparationID,
		PrerequisiteStep:          x.PrerequisiteStep,
		MinEstimatedTimeInSeconds: x.MinEstimatedTimeInSeconds,
		MaxEstimatedTimeInSeconds: x.MaxEstimatedTimeInSeconds,
		TemperatureInCelsius:      x.TemperatureInCelsius,
		Notes:                     x.Notes,
		RecipeID:                  x.RecipeID,
	}
}

File: frontend/src/models/recipeStep.ts
import { RecipeStepIngredient } from "./recipeStepIngredient";

export class RecipeStep {
     id: number;
     temperatureInCelsius?: number;
     externalID: string;
     why: string;
     notes: string;
     index: number;
     prerequisiteStep: number;
     preparationID: number;
     maxEstimatedTimeInSeconds: number;
     minEstimatedTimeInSeconds: number;
     belongsToRecipe: number;
     ingredients: RecipeStepIngredient[];
     createdOn: number;
     lastUpdatedOn?: number;
     archivedOn?: number;

    constructor(
          id: number = 0,
          temperatureInCelsius?: number,
          externalID: string = "",
          why: string = "",
          notes: string = "",
          index: number = 0,
          prerequisiteStep: number = 0,
          preparationID: number = 0,
          maxEstimatedTimeInSeconds: number = 0,
          minEstimatedTimeInSeconds: number = 0,
          belongsToRecipe: number = 0,
          ingredients: RecipeStepIngredient[] = [],
          createdOn: number = 0,
          lastUpdatedOn?: number,
          archivedOn?: number,
    ) {
         this.id = id;
         this.temperatureInCelsius = temperatureInCelsius;
         this.externalID = externalID;
         this.why = why;
         this.notes = notes;
         this.index = index;
         this.prerequisiteStep = prerequisiteStep;
         this.preparationID = preparationID;
         this.maxEstimatedTimeInSeconds = maxEstimatedTimeInSeconds;
         this.minEstimatedTimeInSeconds = minEstimatedTimeInSeconds;
         this.belongsToRecipe = belongsToRecipe;
         this.ingredients = ingredients;
         this.createdOn = createdOn;
         this.lastUpdatedOn = lastUpdatedOn;
         this.archivedOn = archivedOn;
    }
}

export class RecipeStepList {
     recipeSteps: RecipeStep[];
     totalCount: number;
     page: number;
     limit: number;
     filteredCount: number;

     constructor(
          recipeSteps: RecipeStep[] = [],
          totalCount: number = 0,
          page: number = 0,
          limit: number = 0,
          filteredCount: number = 0,
     ) {
          this.recipeSteps = recipeSteps;
          this.totalCount = totalCount;
          this.page = page;
          this.limit = limit;
          this.filteredCount = filteredCount;
     }
}